cpp/include/raft/distance/detail/fused_distance_nn.cuh

```diff
@@ -0,0 +1,89 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cstddef>                                       // size_t
+#include <limits>                                        // std::numeric_limits
+#include <raft/core/kvp.hpp>                             // raft::KeyValuePair
+#include <raft/core/operators.hpp>                       // raft::identity_op
+#include <raft/distance/detail/distance_ops/l2_exp.cuh>  // ops::l2_exp_distance_op
+#include <raft/distance/detail/fused_distance_nn/cutlass_base.cuh>
+#include <raft/distance/detail/fused_distance_nn/fused_cosine_nn.cuh>
+#include <raft/distance/detail/fused_distance_nn/helper_structs.cuh>
+#include <raft/distance/detail/fused_distance_nn/simt_kernel.cuh>
+#include <raft/distance/detail/pairwise_distance_base.cuh>  // PairwiseDistances
+#include <raft/distance/distance_types.hpp>
+#include <raft/linalg/contractions.cuh>  // Policy
+#include <raft/util/arch.cuh>            // raft::util::arch::SM_*
+#include <raft/util/cuda_utils.cuh>      // raft::ceildiv, raft::shfl
+
+namespace raft {
+namespace distance {
+
+namespace detail {
+
+template <typename DataT,
+          typename OutT,
+          typename IdxT,
+          typename Policy,
+          typename ReduceOpT,
+          typename KVPReduceOpT>
+void fusedDistanceNNImpl(OutT* min,
+                         const DataT* x,
+                         const DataT* y,
+                         const DataT* xn,
+                         const DataT* yn,
+                         IdxT m,
+                         IdxT n,
+                         IdxT k,
+                         int* workspace,
+                         ReduceOpT redOp,
+                         KVPReduceOpT pairRedOp,
+                         bool sqrt,
+                         bool initOutBuffer,
+                         bool isRowMajor,
+                         raft::distance::DistanceType metric,
+                         float metric_arg,
+                         cudaStream_t stream)
+{
+  // The kernel policy is determined by fusedDistanceNN.
+  typedef Policy P;
+
+  dim3 blk(P::Nthreads);
+  auto nblks            = raft::ceildiv<int>(m, P::Nthreads);
+  constexpr auto maxVal = std::numeric_limits<DataT>::max();
+  typedef KeyValuePair<IdxT, DataT> KVPair;
+
+  RAFT_CUDA_TRY(cudaMemsetAsync(workspace, 0, sizeof(int) * m, stream));
+  if (initOutBuffer) {
+    initKernel<DataT, OutT, IdxT, ReduceOpT>
+      <<<nblks, P::Nthreads, 0, stream>>>(min, m, maxVal, redOp);
+    RAFT_CUDA_TRY(cudaGetLastError());
+  }
+
+  switch (metric) {
+    case DistanceType::CosineExpanded:
+      fusedCosineNN<DataT, OutT, IdxT, P, ReduceOpT, KVPReduceOpT>(
+        min, x, y, xn, yn, m, n, k, workspace, redOp, pairRedOp, sqrt, stream);
+      break;
+    default: assert("only cosine metric is supported with fusedDistanceNN\n"); break;
```

Comment:
Why don't we dispatch L2 distance?

Reply:
only reason is to avoid multiple entry points for L2 distance and not break the existing fusedL2NN API for downstream RAFT users.
But I think it is better to add it here and later add deprecate the fusedL2NN API always allowing it to be called from fusedDistanceNN instead.

Reply:
@cjnolet currently fused_l2_nn_min_arg supports fp32 & fp64 in `libraft/pylibraft`. 
This PR adds a new cosine fused nn kernel exposed via new API `fused distance nn`  which we plan expose only as fp32 in `libraft/pylibraft`, we want to deprecate/drop `fused_l2_min_arg` in future release in favor of `fused_distance_nn_min_arg` as it provides common API for any distance metric fused with 1-NN feature. For that to happen we shall have to remove its fp64 support, will it be fine to do that?
Or do you think `fused distance nn`  should also expose fp64 support in libraft/pylibraft.

Note that `fused distance nn` does support fp64 we test that it in raft c++ testsuite only we don't expose in libraft/pylibraft. And if dropping fp64 support from `fused_l2_min_arg` is fine that can allow to further unify its code with fused distance nn code base.

Reply:
@cjnolet  *ping* for response on comment above.

Reply:
I believe we can accept the current version of the PR. It is a non-braking change, the question is what data type to support in the newly added `fused_distance_nn` functions. "Header only" we support and test both fp32 and fp64. In the compiled lib (and therefore in the Python API) we only support fp32. If needed, we can still add fp64 instantiations later to the library.